pkg/cmd/cli/cmd/set/triggers.go

```diff
@@ -0,0 +1,744 @@
+		}
+		if len(ref.Registry) > 0 || len(ref.ID) > 0 {
+			return fmt.Errorf("the value of --from-image must point to an image stream tag on this server")
+		}
+		if len(ref.Tag) == 0 {
+			return fmt.Errorf("the value of --from-image must include the tag you wish to pull from")
+		}
+		o.FromImage = ref.NameString()
+		o.FromImageNamespace = defaultNamespace(ref.Namespace, cmdNamespace)
+	}
+
+	count := o.count()
+	o.Reset = count == 0 && (o.Auto || o.Manual)
+	switch {
+	case count == 0 && !o.Remove && !o.RemoveAll && !o.Auto && !o.Manual:
+		o.PrintTable = true
+	case !o.RemoveAll && !o.Auto && !o.Manual:
+		o.Auto = true
+	}
+
+	mapper, typer := f.Object()
+	o.Builder = resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), kapi.Codecs.UniversalDecoder()).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(explicit, o.Filenames...).
+		SelectorParam(o.Selector).
+		ResourceTypeOrNameArgs(o.All, args...).
+		Flatten()
+
+	output := kcmdutil.GetFlagString(cmd, "output")
+	if len(output) != 0 {
+		o.PrintObject = func(obj runtime.Object) error { return f.PrintObject(cmd, obj, o.Out) }
+	}
+
+	o.Encoder = f.JSONEncoder()
+	o.ShortOutput = kcmdutil.GetFlagString(cmd, "output") == "name"
+	o.Mapper = mapper
+
+	return nil
+}
+
+func (o *TriggersOptions) count() int {
+	count := 0
+	if o.FromConfig {
+		count++
+	}
+	if o.FromGitHub != nil {
+		count++
+	}
+	if o.FromWebHook != nil {
+		count++
+	}
+	if len(o.FromImage) > 0 {
+		count++
+	}
+	return count
+}
+
+func (o *TriggersOptions) Validate() error {
+	count := o.count()
+	switch {
+	case o.Auto && o.Manual:
+		return fmt.Errorf("you must specify at most one of --auto or --manual")
+	case o.Remove && o.RemoveAll:
+		return fmt.Errorf("you must specify either --remove or --remove-all")
+	case o.RemoveAll && (count != 0 || o.Auto || o.Manual):
+		return fmt.Errorf("--remove-all may not be used with any other flag")
+	case o.Remove && count < 1:
+		return fmt.Errorf("--remove requires a flag defining a trigger type to be specified")
+	case count > 1:
+		return fmt.Errorf("you may only set one trigger type at a time")
+	case count == 0 && !o.Remove && !o.RemoveAll && !o.Auto && !o.Manual && !o.PrintTable:
+		return fmt.Errorf("specify one of the --from-* flags to add a trigger, --remove to remove, or --auto|--manual to control existing triggers")
+	}
+	return nil
+}
+
+func (o *TriggersOptions) Run() error {
+	infos := o.Infos
+	singular := len(o.Infos) <= 1
+	if o.Builder != nil {
+		loaded, err := o.Builder.Do().IntoSingular(&singular).Infos()
+		if err != nil {
+			return err
+		}
+		infos = loaded
+	}
+
+	if o.PrintTable && o.PrintObject == nil {
+		return o.printTriggers(infos)
+	}
+
+	patches := CalculatePatches(infos, o.Encoder, func(info *resource.Info) (bool, error) {
+		return UpdateTriggersForObject(info.Object, func(triggers *TriggerDefinition) error {
+			o.updateTriggers(triggers)
+			return nil
+		})
+	})
```

Comment:
The wrinkle is that the initial object is calculated (trigger),
mutated (fn) and then applied back to the object, with the result
being two objects that can be diffed.  So some nesting is fundamental,
although if you can think of a better factoring I'll apply it.